src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -0,0 +1,104 @@
+#include <cassert>
+#include <algorithm>
+#include <iostream>
+#include "bf_batch_iterator.h"
+#include "VecSim/utils/arr_cpp.h"
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/query_result_struct.h"
+
+unsigned char BF_BatchIterator::next_id = 0;
+
+VecSimQueryResult *BF_BatchIterator::heapBasedSearch(size_t n_res) {
+    float upperBound = std::numeric_limits<float>::lowest();
+    CandidatesHeap TopCandidates;
+    vector<VectorBlock *> blocks = getIndex()->getVectorBlocks();
+    for (size_t i = 0; i < blocks.size(); i++) {
+        blocks[i]->heapBasedSearch(scores[i], upperBound, n_res, TopCandidates);
+    }
+
+    auto *results = array_new_len<VecSimQueryResult>(TopCandidates.size(), TopCandidates.size());
+    for (int i = (int)TopCandidates.size() - 1; i >= 0; --i) {
+        VecSimQueryResult_SetId(results[i], TopCandidates.top().second);
+        VecSimQueryResult_SetScore(results[i], TopCandidates.top().first);
+        // Erase vector score, so we won't return it again in the next iterations.
+        pair<size_t, size_t> coordinates = labelToScoreCoordinates[TopCandidates.top().second];
+        scores[coordinates.first].erase(scores[coordinates.first].begin() +
+                                        (int)coordinates.second);
+        TopCandidates.pop();
+    }
+    return results;
+}
+
+VecSimQueryResult *BF_BatchIterator::selectBasedSearch(size_t n_res) {
+    // Flatten the score matrix, keep the labels to have stable sorting.
+    vector<pair<float, labelType>> allScores;
+    for (auto vectorBlockScores : scores) {
+        allScores.insert(allScores.end(), vectorBlockScores.begin(), vectorBlockScores.end());
+    }
+
+    auto n_th_element_pos = allScores.begin() + (int)n_res;
+    std::nth_element(allScores.begin(), n_th_element_pos, allScores.end());
+
+    auto *results = array_new<VecSimQueryResult>(n_res);
+    for (size_t i = 0; i < n_res; i++) {
+        results = array_append(results, VecSimQueryResult{});
+        VecSimQueryResult_SetId(results[array_len(results) - 1], allScores[i].second);
+        VecSimQueryResult_SetScore(results[array_len(results) - 1], allScores[i].first);
+        // Erase vector score, so we won't return it again in the next iterations.
+        pair<size_t, size_t> coordinates = labelToScoreCoordinates[allScores[i].second];
+        scores[coordinates.first].erase(scores[coordinates.first].begin() +
+                                        (int)coordinates.second);
+    }
+    return results;
+}
+
+BF_BatchIterator::BF_BatchIterator(const void *query_vector, const BruteForceIndex *bf_index)
+    : VecSimBatchIterator(query_vector), id(BF_BatchIterator::next_id) {
+    index = bf_index;
+    BF_BatchIterator::next_id++;
```

Comment:
can you do all of it in the initialization list?
also, see if you can pre-allocate `scores` to the right size (avoid re allocation while appendeing)

Reply:
Added `scores.reserve(index_size)` call in the first iteration of `getNextResults`